### ISSUE TITLE
Auto completion for `Literal` types with long string inhabitants is truncated.

### ISSUE THREAD

**Issue by m0rphism** [microsoft/pyright#11075]
Consider a `Literal` type with a very long string inhabitant, e.g.:
```python
from typing import Literal

Foo = Literal["FoooooooooooooooooooooooooooooooooooooooooooooooooooooooooX"]
```

If autocompletion for such a type is triggered, e.g. in
```python
x: Foo = "
```

then a truncated version of the string is both displayed and inserted, i.e.
```python
x: Foo = "Fooooooooooooooooooooooooooooooooooooooooooooooooo…"
```

The expected behavior would be that autocompletion inserts the original, untruncated value, i.e.
```python
x: Foo = "FoooooooooooooooooooooooooooooooooooooooooooooooooooooooooX"
```

The behavior was detected in neovim by using `pyright` as an lsp-server and `cmp` as an autocompletion plugin using the `nvim_lsp` source.

The issue seems to be not related to the completion plugin `cmp`, as it affects exclusively completions for values of a `Literal` type, and other completions, e.g. for a variable with a very long identifier, work as expected. Furthermore, the ellipsis is inserted inside of the string ( `…"`), which seems to suggest that the truncation happens inside of pyright.

**Comment by erictraut** [comment]
Yes, this is by design. Pyright internally truncates literal str and bytes types because string can be arbitrarily long in Python. If a string is many megabytes in length, it's unlikely that you want to auto-insert the full string as a literal. There has to be a cutoff at some point.

**Comment by m0rphism** [comment]
Thanks for the quick reply.

> If a string is many megabytes in length, it's unlikely that you want to auto-insert the full string as a literal.

While I completely agree that string values with megabytes of size are not reasonable for a `Literal` type, and support can be dropped in favor of guaranteeing responsive tooling, I would argue that the current limit is too small and needlessly excludes rare but useful scenarios.

For example, in my current use case I auto-generate `Literal` types that describe hierarchical device ids.
These device ids have the same shape as file-system paths, can contain spaces, and are often around 300 characters long.
Due to the special characters in the ids, I would need some id mangling to map them to actual python identifiers, which would be ugly and has the potential of collisions or nondeterminism. Hence, I generate a `Literal` type for them, such that I can simply write them down verbatim, get type safety and auto-completion.

While this is admittedly not a very common use cases, increasing the truncation limit to a few kilobytes would easily support this and probably most related use cases without introducing responsiveness issues as they might occur from huge string constants of multiple megabytes or larger and any decrease in responsiveness would only occur in those niche projects where large string constants are actually used.

Basically my point is: while a 1000-line constant in a literal type is obviously nonsense, a 3-line constant can be useful in niche cases, and there seems to be no drawback in supporting them.

**Comment by erictraut** [comment]
I don't think that increasing the truncation limit would serve most pyright or pylance users well. It would affect readability of hover text. 

I'm also skeptical of your use case. I wouldn't recommend using `Literal` to define types with 300+ characters.

**Comment by m0rphism** [comment]
> I don't think that increasing the truncation limit would serve most pyright or pylance users well. It would affect readability of hover text.

Wouldn't it be possible to display them truncated but to insert them without truncation on auto-completion?

**Comment by Minibrams** [comment]
From [another discussion](https://github.com/microsoft/pylance-release/issues/5116#issuecomment-3291093083) on the Pylance repo, but same issue:

@erictraut Any chance this could be revisited?

I understand that adding knobs for controlling max literal length seems like a strange thing to add, but there are scenarios where it makes sense. I don't think these are particularly esoteric.

I am currently using a library that auto-generates API clients from OpenAPI specs, and it uses string literals to ensure type-safety when calling endpoints - just like [openapi-typescript](https://github.com/openapi-ts/openapi-typescript) does, among others. Some of these endpoints are long, but not unreasonably so:

```python
# Actual endpoint is: "/api/v1/git/merge_request/{config_type}/{site}/{name}"
ApiClient().get_route("/api/v1/git/merge_request/{config_type}/{site}/{na…")
```

Another scenario is i18n where string literals are used for translation keys:
```python
raise NotFoundError(i18n.get("customer.customer_db_config_page.config_not_fou…"))
```

I don't know if these count as esoteric, but it does make it really hard to use string literals for stuff like this. I have to jump into the source code each time to get the full string literal every time it's too long for autocompletion.

The length is hardcoded [here](https://github.com/microsoft/pyright/blob/5f7ce347e8fbf80c826870b4a95395916cce3454/packages/pyright-internal/src/analyzer/typePrinter.ts#L138) - could it be considered to use 50 as a default, but let the user override it with a configuration property e.g. `maxLiteralStringLength`? If so, I'll gladly create a PR for it.
